torch/cuda/amp/grad_scaler.py

```diff
@@ -180,25 +180,36 @@ def _unscale_grads_(self, optimizer, inv_scale, found_inf, allow_fp16):
         per_device_inv_scale = _MultiDeviceReplicator(inv_scale)
         per_device_found_inf = _MultiDeviceReplicator(found_inf)
 
+        # To set up _amp_foreach_non_finite_check_and_unscale_, split grads by device and dtype.
+        # There could be hundreds of grads, so we'd like to iterate through them just once.
+        # However, we don't know their devices or dtypes in advance.
+        per_device_and_dtype_grads = defaultdict(defaultdict(list))  # https://knowyourmeme.com/memes/xzibit-yo-dawg
         for group in optimizer.param_groups:
             for param in group["params"]:
                 if param.grad is not None:
                     if (not allow_fp16) and param.grad.dtype == torch.float16:
                         raise ValueError("Attempting to unscale FP16 gradients.")
                     else:
-                        with torch.no_grad():
-                            if param.grad.is_sparse:
-                                # is_coalesced() == False means the sparse grad has values with duplicate indices.
-                                # coalesce() deduplicates indices and adds all values that have the same index.
-                                # For scaled fp16 values, there's a good chance coalescing will cause overflow,
-                                # so we should check the coalesced _values().
-                                if param.grad.dtype is torch.float16:
-                                    param.grad = param.grad.coalesce()
-                                to_unscale = param.grad._values()
-                            else:
-                                to_unscale = param.grad
-
-                            torch._amp_non_finite_check_and_unscale_(to_unscale,
+                        if param.grad.is_sparse:
+                            # is_coalesced() == False means the sparse grad has values with duplicate indices.
+                            # coalesce() deduplicates indices and adds all values that have the same index.
+                            # For scaled fp16 values, there's a good chance coalescing will cause overflow,
+                            # so we should check the coalesced _values().
+                            if param.grad.dtype is torch.float16:
+                                param.grad = param.grad.coalesce()
+                            to_unscale = param.grad._values()
+                        else:
+                            to_unscale = param.grad
+
+                        # I don't like appending in the inner loop, but I can't think of a better way
+                        # to populate per-device-and-dtype lists in one pass where devices and dtypes
+                        # are not known in advance.
```

Comment:
nit: i think this comment can be removed.

Reply:
Aye, it was for reviewers, if anyone can thing of a more Pythonic way to iterate through all the params and organize them into lists by device and dtype in one shot, please tell me...I could use a generator expression + map but I don't think that would save me from appending to lists.

Reply:
Shortened to `# TODO: is there a way to split by device and dtype without appending in the inner loop?`
I want to keep a reminder in case some reader comes up with a genius python hack.